backend-api/src/functions/adapters/dynamoDbAdapter.ts

```diff
@@ -0,0 +1,195 @@
+      Limit: 1,
+      ScanIndexForward: false,
+    };
+
+    const queryResponse = await this.query(input);
+    if (queryResponse.isError || queryResponse.value === null) {
+      return queryResponse;
+    }
+
+    const record: DynamoDbRecord = queryResponse.value;
+    const sessionId = record.sessionId?.S;
+    const state = record.state?.S;
+    if (!sessionId || !state) {
+      return errorResult({
+        errorMessage: "Session is malformed",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+    const redirectUri: string | undefined = record.redirectUri?.S;
+    const sessionDetails: SessionDetails = {
+      sessionId,
+      state,
+      ...(redirectUri && { redirectUri: record.redirectUri?.S }),
+    };
+
+    return successResult(sessionDetails);
+  }
+
+  async createSession(
+    attributes: CreateSessionAttributes,
+    sessionId: string,
+  ): Promise<Result<string>> {
+    const {
+      client_id,
+      govuk_signin_journey_id,
+      issuer,
+      redirect_uri,
+      sessionDurationInSeconds,
+      state,
+      sub,
+    } = attributes;
+    const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+    const timeToLive = currentTimeInSeconds + sessionDurationInSeconds;
+
+    const input: PutItemCommandInput = {
+      TableName: this.tableName,
+      Item: {
+        clientId: { S: client_id },
+        govukSigninJourneyId: { S: govuk_signin_journey_id },
+        createdAt: { N: Date.now().toString() },
+        issuer: { S: issuer },
+        sessionId: { S: sessionId },
+        sessionState: { S: "ASYNC_AUTH_SESSION_CREATED" },
+        state: { S: state },
+        subjectIdentifier: { S: sub },
+        timeToLive: { N: timeToLive.toString() },
+      },
+      ConditionExpression: "attribute_not_exists(sessionId)",
+    };
+
+    if (redirect_uri) {
+      input.Item!.redirectUri = { S: redirect_uri };
+    }
+
+    try {
+      await this.dynamoDbClient.send(new PutItemCommand(input));
+    } catch (error) {
+      if (error instanceof ConditionalCheckFailedException) {
+        return errorResult({
+          errorMessage: "Session already exists with this ID",
+          errorCategory: "SERVER_ERROR",
+        });
+      } else {
+        return errorResult({
+          errorMessage: "Unexpected error while creating a new session",
+          errorCategory: "SERVER_ERROR",
+        });
+      }
+    }
+
+    return successResult(sessionId);
+  }
+
+  private async query(
+    input: QueryCommandInput,
+  ): Promise<Result<DynamoDbRecord> | Result<null>> {
+    let output: QueryCommandOutput;
+    try {
+      output = await this.dynamoDbClient.send(new QueryCommand(input));
+    } catch {
+      return errorResult({
+        errorMessage: "Unexpected error when querying database",
+        errorCategory: "SERVER_ERROR",
+      });
+    }
+
+    if (!output.Items || output.Items.length === 0) {
+      return successResult(null);
+    }
+
+    return successResult(output.Items[0]);
```

Comment:
what happens if there are two items in the array - or is this scenario not expected? This feels like logic that should live in the function that returned. This function is now doing two things - making the query and choosing to return the first item

Reply:
You are right. The two query inputs we have right now do not allow two items to be returned because they both specify one item only, so taking into account how we are using this `query` method now, returning the first item is not a problem. However, you are absolutely right that it should not return output.Items[0] even if it can - it's not clear at all and makes it impossible to re-use the method with a different type of query input. I will fix that... 👀

Reply:
Fixed and method no longer exists